### PR TITLE
CLN: Fixing mypy errors in pandas/conftest.py

### PR DESCRIPTION
**Pull request by angelaambroz**
- [x] xref #28926 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] ~whatsnew entry~

### CODE REVIEW

**Comment by WillAyd** [comment]
Can you annotate these instead of ignoring? I think should can just use Dtype from pandas._typing. Might need to modify that to account for built in types

**Comment by angelaambroz** [comment]
Cool - hadn't seen pandas._typing. Will do.

**Comment by angelaambroz** [comment]
I've left the `# type: ignore` for the import statement, since I suspect it's related to [what the mypy docs describe](https://mypy.readthedocs.io/en/latest/running_mypy.html#how-mypy-handles-imports) when a module is present, but mypy is unable - for whatever reason - to check its type hints.

File: pandas/conftest.py
Comment (WillAyd) [inline pandas/conftest.py]: This is actually a `List[Dtype]`

File: pandas/conftest.py
Comment (WillAyd) [inline pandas/conftest.py]: This is the same as #28932 - I think this needs to be fixed upstream in Typeshed. Would you be interested in submitting that there?

File: pandas/conftest.py
Comment (WillAyd) [reply pandas/conftest.py]: I think `Dtype` needs to be updated in pandas._typing to include `Type[Union[float, int]]` for when the built-in values get provided - could you make that change as well?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/conftest.py
Comment (angelaambroz) [reply pandas/conftest.py]: Sure - I can give it a shot.

File: pandas/conftest.py
Comment (angelaambroz) [reply pandas/conftest.py]: Yep, sure - thanks for the guidance.

Comment (simonjayhawkins) [reply pandas/conftest.py]: `Type[Union[float, str, bool]]` would be sufficient as int is duck type compatible with float see https://mypy.readthedocs.io/en/latest/duck_type_compatibility.html#duck-type-compatibility

Comment (angelaambroz) [reply pandas/conftest.py]: Thanks, both. Just pushed what I _think_ you both meant - let me know. I'm still learning about types! Looks like `complex` was also duck typed to `float`.

**Comment by WillAyd** [comment]
You should encapsulate those in Type[...] otherwise this would be looking for an instance of those. 

FWIW I’d rather be explicit about these rather than relying on any hierarchy or inference. In the future we’d probably encapsulate in Literal

File: pandas/_typing.py
Comment (simonjayhawkins) [inline pandas/_typing.py]: need to import Type from typing

Comment (angelaambroz) [reply pandas/_typing.py]: 🤦‍♀ Sorry about that. Fixing that now.

Comment (WillAyd) [inline pandas/_typing.py]:
```suggestion
Dtype = Union[ExtensionDtype, Type[Union[str, float, int, complex, bool]]]
```

Somewhat more complicated but the `ExtensionDtype` needs to be an instance, so should exclude from the `Type` element.

To illustrate:

```python
>>> pd.Series(range(1), dtype=pd.Int64Dtype)  # Not what you would expect
0    0
dtype: object
>>> pd.Series(range(1), dtype=pd.Int64Dtype())  # This instance works though
0    0
dtype: object
>>> pd.Series(range(1), dtype=int) # contrast to the builtin, which uses Type[int]
0    0
dtype: int64
```

cc @TomAugspurger in case there's something I'm missing with the ExtensionDtype

Also @simonjayhawkins do you mind if we include `int` in the Union? Understood that it gets replaced as part of the mypy type hierarchy, but given ``dtype=int`` is a valid construct it serves as literal documentation and requires less knowledge of how mypy works. Whenever we get to Python 3.8 as the min version these *should* be converted to `Literal` annotations anyway

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_typing.py
Comment (TomAugspurger) [reply pandas/_typing.py]:
```
>>> pd.Series(range(1), dtype=pd.Int64Dtype)
```

should probably raise. I don't think we can automatically make that work since we have parametrized dtypes. Not sure what that does to the type hint.

Comment (simonjayhawkins) [reply pandas/_typing.py]: > Also @simonjayhawkins do you mind if we include `int` in the Union?

no problem.

File: pandas/conftest.py
Comment (angelaambroz) [reply pandas/conftest.py]: Picking this up again - just got the changes merged in typeshed (https://github.com/python/typeshed/pull/3393). I see that there are now some conflicts with master; I can address those and pick this up again in the next couple days.

**Comment by alimcmaster1** [comment]
@angelaambroz thanks for the PR! Are you still working on this?

**Comment by angelaambroz** [comment]
Eek, sorry, yes. I'd like to keep working on this. Apologies for the silence.

File: pandas/conftest.py
Comment (simonjayhawkins) [inline pandas/conftest.py]: will need to change to py3.6 syntax.

File: pandas/conftest.py
Comment (simonjayhawkins) [inline pandas/conftest.py]: same

File: pandas/_typing.py
Comment (angelaambroz) [reply pandas/_typing.py]: 👋 Sorry for the silence on my end. I'm picking this up again now. I didn't follow your comment, @WillAyd (though I committed your suggestion - hence the `Outdated` flag). Does this requested review still apply? Thanks.

**Comment by angelaambroz** [comment]
CI fails on `pytest pandas/conftest.py`, due to `ExtensionDtype` not being recognized. It's currently only imported if `TYPE_CHECKING==True`. There's a comment warning against circular imports, so I'm hesitating moving
```
from pandas.core.dtypes.dtypes import ExtensionDtype  # noqa: F401
```
out of the conditional. Halp? 🆘

File: pandas/_typing.py
Comment (WillAyd) [inline pandas/_typing.py]:
```suggestion
Dtype = Union["ExtensionDtype", Type[Union[str, float, int, complex, bool]]]
```

Should be quoted. My mistake on earlier suggestion

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by datapythonista** [comment]
@angelaambroz can you take care of the last comment, so we can finally merge this please? Please ping us once the CI is green, for a faster merge.

**Comment by angelaambroz** [comment]
Argh, sorry, I'm traveling and away from a laptop until Jan 8. Can someone
else do the last push? Sorry it's been hanging so long.

**Comment by datapythonista** [comment]
You can do the last commit from the website, there is a dropdown to accept the proposed change. I tried to do it myself, but looks like GitHub just let the author accept it.

**Comment by angelaambroz** [comment]
Ack. CI failing linting - it doesn't like the import order in conftest. 
- I could try fixing this via the GitHub web UI (iffy, but I might be able to do it).
- Someone else picks this up.
- We wait until Jan 8/9?

**Comment by alimcmaster1** [comment]
> Ack. CI failing linting - it doesn't like the import order in conftest.
> 
> * I could try fixing this via the GitHub web UI (iffy, but I might be able to do it).
> * Someone else picks this up.
> * We wait until Jan 8/9, when I'm back at a computer and can debug properly.

You can run `isort -rc pandas` to fix the import order issue. For more info see here: https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#import-formatting

**Comment by WillAyd** [comment]
@angelaambroz can you fix up merge conflict? I think this is pretty much ready

**Comment by angelaambroz** [comment]
It's failing on type checks now; this was unexpected, but I didn't dive deep into the merge conflicts on `pandas/_typing.py`, where I saw some changes (I intended just to change the `Dtype` type). I'll try to go through this later today and see what's up.

File: pandas/_typing.py
Comment (WillAyd) [inline pandas/_typing.py]:
```suggestion
Dtype = Union["ExtensionDtype", str, Type[Union[str, float, int, complex, bool]]]
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
I think just need to add `str` as a valid dtype to resolve CI errors. Something like `dtype="boolean"` is valid

**Comment by simonjayhawkins** [comment]
@angelaambroz can you merge master

**Comment by MarcoGorelli** [comment]
Hi @angelaambroz  - sorry to chase you up, just wanted to ask whether you're still working on this :)

**Comment by angelaambroz** [comment]
@MarcoGorelli Feel free to grab it - I'm being very slow and not finding enough time to get back to this! I'd be happy to continue, but don't want to hold things up. Thanks for checking in!

**Comment by MarcoGorelli** [comment]
> I'd be happy to continue, but don't want to hold things up

It's OK, I don't believe this is _too_ urgent, so feel free to continue and to ask for help if you're stuck

**Comment by WillAyd** [comment]
@angelaambroz any interest in fixing up the merge conflict and addressing last comment?

**Comment by angelaambroz** [comment]
👋 Yes, here! I hang my head in shame. But I can do this later today.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @angelaambroz for seeing this through

**Pull request merged**
